### PR TITLE
Bug 1519995 - Perfherder graphs react conversion part 3

### PR DESCRIPTION
**Pull request by sarah-clements**
So this is finally done and has been launched to prototype :) Cam, you only need to look at the second commit because the first was reviewed in #5158 .

I've managed to replicate all of the jquery graph library functionality with the new Victory (react component based) library. There were two things I wasn't able to implement successfully due to bugs with the graph library (I've filed issues upstream): highlighting the data point when it's been clicked/selected and using a selection container on the main lower graph (it's been replaced with mouse wheel zoom). Once those bugs are fixed, you can implement that functionality if desired.

There's a few things, as mentioned in the last Perfherder meeting, that I didn't get to - mainly deprecating the use of signature hash's in the url and a few existing bugs for general improvements.

Regarding tests, I've replaced the selenium tests I had disabled with react-testing library tests. Since many changes would've had to be made, I thought it best to start fresh and make it easier for people to add more tests in the future (rather than having two different libraries implement the same type of testing). I'd like to add tests for the tooltip but I haven't finished them yet.

@ionutgoldan please test the library on prototype. If you need more time to test it and someone else has used prototype, please ask Cam or Armen to push it to prototype again. Something I'd like feedback on is the max number of tests to support showing for optimal performance and visibility (either through messaging or by limiting the number of legend colors which is currently at 6). Six might be ok with short time ranges but for longer time ranges I think performance would be slow when using mouse zoom.

FYI @karlht and @davehunt

### CODE REVIEW

File: ui/perfherder/App.jsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                />
+              )}
+            />
+            <Route path="/graphs" component={GraphsView} />
+            <Route
+              path="/graphs?timerange=:timerange&series=:series&highlightedRevisions=:highlightedRevisions&highlightAlerts=:highlightAlerts&zoom=:zoom&selected=:selected"
+              component={GraphsView}
+            />
+            <Route path="/compare" component={CompareView} />
+            {/* <Route
```
Comment (camd) [inline ui/perfherder/App.jsx]: Is this for something coming up in a future PR?

File: ui/perfherder/graphs/GraphsContainer.jsx [hunk 2]
```diff
@@ -1,15 +1,484 @@
...
+    const {
+      highlights,
+      scatterPlotData,
+      entireDomain,
+      showTooltip,
+      lockTooltip,
+      dataPoint,
+    } = this.state;
+
+    const highlightPoints = Boolean(highlights.length);
```
Comment (camd) [inline ui/perfherder/graphs/GraphsContainer.jsx]: You can also use a shortcut for things like this: ``!!highlights.length`` which will turn it into a boolean.

File: ui/perfherder/index.jsx [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import { render } from 'react-dom';
+import 'bootstrap/dist/css/bootstrap.min.css';
+import 'react-table/react-table.css';
+
+import '../css/treeherder-global.css';
+// import '../css/treeherder-navbar.css';
```
Comment (camd) [inline ui/perfherder/index.jsx]: should this line be removed?

File: ui/css/perf.css [hunk 4]
```diff
@@ -87,54 +69,56 @@ h1 {
   width: 120px;
 }
 
-#graph-tooltip {
+.graph-tooltip {
   position: absolute;
-  visibility: hidden;
+  /* visibility: hidden; */
```
[the remark this answers is not in the document]
Comment (sarah-clements) [reply ui/css/perf.css]: yes it should :)

File: ui/perfherder/App.jsx (hunk 1 above)
Comment (sarah-clements) [reply ui/perfherder/App.jsx]: oops, it is... so strange. Let me remove that stuff :) Clearly I'm ready for a vacation!

File: ui/css/perf.css [hunk 5]
```diff
@@ -87,54 +69,56 @@ h1 {
...
-#graph-tooltip.locked .tip {
+.graph-tooltip.locked .tip {
   background-image: url('../img/tip-locked.png');
 }
 
 .graphchooser-close {
   color: #fff;
   outline: none;
-  margin: -14px -4px;
+  /* margin: -14px -4px; */
```
[the remark this answers is not in the document]
Comment (sarah-clements) [reply ui/css/perf.css]: done

File: ui/perfherder/App.jsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import { CompareView } from './compare/CompareView';
+
+class App extends React.Component {
+  constructor(props) {
+    super(props);
+
+    // keep track of the mainviews graph and table data so the API won't be
+    // called again when navigating back from bugdetailsview.
+    this.state = {
+      // graphData: null,
```
[the remark this answers is not in the document]
Comment (sarah-clements) [reply ui/perfherder/App.jsx]: I'll remove this file and the index... committed by mistake.

**Comment by ionutgoldan** [comment]
This PR looks very promising! With some slight adjustments & tweaks, it'll be good to land.

**Comment by ionutgoldan** [comment]
I'm proving the STR to the bugs I've encountered in the PR. They look to be pretty small.

**#1** Inconsistency on alerts' highlighting
**STR**
1. Go to a random plotted signature ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/graphs?timerange=1209600&series=autoland,1922287,1,1&highlightAlerts=1&zoom=1565601203000,1566796005000,5.935358101820672,9.33365078888615))
2. Highlight a random data point, which hasn't got any signature (e.g. hash c789725af1e47)
3. Click on ""Highlight alerts"

Expected results:
Data points associated to alerts have empty circles & aren't highlighted in any way.

What we actually get:
These data points have full circles & are highlighted by a bigger circle.

**Comment by ionutgoldan** [comment]
@sarah-clements could you make both graphs use the right side, like in the older version?
Also, it'd be better if the upper graph would have a bigger height (+20%).

**Comment by ionutgoldan** [comment]
Another request? Is it possible to make the mouse scroll on the lower graph to zoom in exponentially?
That way, the investigation process would be significantly faster.

**Comment by ionutgoldan** [comment]
**#2** Repository hyperlink broken

**STR**
1. Go to a random plotted signature ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/graphs?timerange=31536000&series=autoland,1922287,1,1&highlightAlerts=1&zoom=1553618214000,1566800644000,2.2663364433986772,11.058896082814988))
2. Click on the **autoland** hyperlink from top-left *(between the test name and the platform name)*.

Expected results:
**Related tests** should show a single test with the same name & from the same platform.

What we get instead:
Lots of tests, with multiple names & from multiple platforms.

**Comment by ionutgoldan** [comment]
Bug similar to **#2**.

**#3** Platform hyperlink broken
**STR**
1. Go to a random plotted signature ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/graphs?timerange=31536000&series=autoland,1922287,1,1&highlightAlerts=1&zoom=1553618214000,1566800644000,2.2663364433986772,11.058896082814988))

**Comment by ionutgoldan** [comment]
**#4** Y-axis numerical legend values are chipped

**STR**
1. Go to Graphs view with data points having high numerical values ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/graphs?timerange=1209600&series=mozilla-inbound,1949187,1,1&series=autoland,a4b827f053f43d02a4ee20f55ab1cbcf227c7431,1,1))

Expected results:
Numerical legend values are correct & readable.

What we get instead: 
Numerical values having more than 5 characters in length aren't fully visible; they're first digit is chipped.

**Comment by ionutgoldan** [comment]
>Something I'd like feedback on is the max number of tests to support showing for optimal performance and visibility (either through messaging or by limiting the number of legend colors which is currently at 6). Six might be ok with short time ranges but for longer time ranges I think performance would be slow when using mouse zoom.

I'd say we keep with the old behavior. An initial warning dialog when user loads more than 6 plots should suffice. A nice feature would be to make visible only the first 6 plots & leave the rest of them hidden. If the user wants to make all visible, it would have to go through each of them hidden ones & check them.

**Comment by sarah-clements** [comment]
> Another request: could the mouse zoom in increase exponentially? In current form, Perf sheriffs would have to do a lot of scroll down/up to navigate within the graph. An exponential zoom in would give benefits from both worlds: we retain precision when we scroll once/twice, we quickly get big navigation speed if scroll more than that.
> 
> This way, the investigation process would be significantly faster.

Hrm, I don't see that feature as part of the graph library. I think it'd take a bit of work to figure out how to do this nicely with how the zoom container functions (it might look a bit jumpy to manually reset the coordinates during the zoom action). Can you could file a bug for it? There seems to be some traction on fixing the onClick bug for the SelectionContainer so another option once it's fixed (https://github.com/FormidableLabs/victory/issues/1367) would be to change the bottom zoom container to that. In the meantime, I'd suggest using the top selection marquee to narrow down an area before zooming with the mouse.

**Comment by sarah-clements** [comment]
> I'd say we keep with the old behavior. That is: an initial warning dialog when user loads more than 6 plots should suffice. A nice feature would be to make visible only the first 6 plots & leave the rest of them hidden. If the user wants to make all visible, he would have to go through each of them hidden ones & check them.

Ah, so I did something like that - I forgot to mention it. Adding more than 6 tests makes them grey and deselected (trying to select it will show a message prompting the user to first deselect another test, so only 6 can be shown at a time). If you find issues with this, I can follow up with a separate pr.

**Comment by sarah-clements** [comment]
> I'm providing the STR to the bugs I've encountered in the PR. They look to be pretty small.
> 
> **#1** Inconsistency on alerts' highlighting
> **STR**
> 
>     1. Go to a random plotted signature ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/graphs?timerange=1209600&series=autoland,1922287,1,1&highlightAlerts=1&zoom=1565601203000,1566796005000,5.935358101820672,9.33365078888615))
> 
>     2. Highlight a random data point, which hasn't got any signature (e.g. hash c789725af1e47)
> 
>     3. Click on "Highlight alerts"
> 
> 
> Expected results:
> Data points associated to alerts have empty circles & aren't highlighted in any way.
> 
> What we actually get:
> These data points have full circles & are highlighted by a bigger circle.

Are you saying that toggling "highlight alerts" doesn't work correctly? It seems to be working fine for me, regardless of whether a data point has been selected previously or not.

**Comment by sarah-clements** [comment]
> **#4** Y-axis numerical legend values are chipped
> 
> **STR**
> 
>     1. Go to Graphs view with data points having high numerical values ([e.g.](https://treeherder-prototype.herokuapp.com/perf.html#/graphs?timerange=1209600&series=mozilla-inbound,1949187,1,1&series=autoland,a4b827f053f43d02a4ee20f55ab1cbcf227c7431,1,1))
> 
> 
> Expected results:
> Numerical legend values are correct & readable.
> 
> What we get instead:
> Numerical values having more than 5 characters in length aren't fully visible; they're first digit is chipped.

I'm not seeing clipped numbers. The scale is different, but they look fine to me.

Screen Shot 2019-09-05 at 5 21 20 PM

**Comment by sarah-clements** [comment]
> @sarah-clements could you make both graphs use the right side, like in the older version?
> Also, it'd be better if the upper graph would have a bigger height (+20%).

I've made both of the graphs a little bigger height and width-wise, however a limitation with svg-based charts vs use of canvas (which the jquery plot graph uses) pertains to the use of aspect ratio. If I make the graphs responsive and don't constrain the parent container, the graph will be quite large height and width wise at very large view ports (!1700px) in order to maintain the aspect ratio. If I remove the aspect ratio so only the width changes depending on the view port, the x axis labels are stretched, which looks pretty bad. The compromise I've found is to make the width a little bigger, but will max out at >= 1700px view port. However, the graphs will still be responsive and scale proportionally for smaller view ports which wasn't the case with my initial commits (so the graphs won't wrap to the next row if you're viewing them from a laptop).

I've fixed the other bugs you've found so I think this is ready to go :)  Thanks for your help with the testing!

**Comment by marianrai** [comment]
Login issue on Firefox and Chrome;

STR
    1. Go to https://treeherder-prototype.herokuapp.com/perf.html
    2. Logged in successfully
    3. After step2 i see a blue alert info on the top : "You must be logged into perfherder/treeherder and be a sheriff to make changes". On top right i can see my name, but i'm not really logged in.
    4. I can't use the checkboxes on alert summaries nor open new alerts from graphs.
    5. Cleared cache, cookies, restarted browsers - issue persists.

Expected results:
After the login i should to have full functionality of perfherder features.

What we get instead:
Perfherder acts as i am not logged in.
In Chrome i can see this error in console:
"Login issuie in heroku perfherder
AuthService.js:66 Could not renew login: Object
_renewAuth @ AuthService.js:66"

**Pull request merged**